#### Banele Blessing Zondo, assignment 2

## Various Python imports

In [ ]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.model_selection import train_test_split

## Load the dataset

In [ ]:
pima = pd.read_csv("regressiondata.csv",delimiter=";")
# Shaun: Good ✅✅

## Read in the data

In [ ]:
pima.head() # Shaun: Good ✅

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0            7.4              0.70         0.00  ...       0.56      9.4        5
1            7.8              0.88         0.00  ...       0.68      9.8        5
2            7.8              0.76         0.04  ...       0.65      9.8        5
3           11.2              0.28         0.56  ...       0.58      9.8        6
4            7.4              0.70         0.00  ...       0.56      9.4        5

[5 rows x 12 columns]

In [ ]:
pima.describe()

fixed acidity  volatile acidity  ...      alcohol      quality
count    1599.000000       1599.000000  ...  1599.000000  1599.000000
mean        8.319637          0.527821  ...    10.422983     5.636023
std         1.741096          0.179060  ...     1.065668     0.807569
min         4.600000          0.120000  ...     8.400000     3.000000
25%         7.100000          0.390000  ...     9.500000     5.000000
50%         7.900000          0.520000  ...    10.200000     6.000000
75%         9.200000          0.640000  ...    11.100000     6.000000
max        15.900000          1.580000  ...    14.900000     8.000000

[8 rows x 12 columns]

In [ ]:
len(pima)

1599

In [ ]:
pima.shape

(1599, 12)

## Create X and y variables (features and target)

In [ ]:
feature_cols =['fixed acidity','volatile acidity','citric acid','residual sugar', 'chlorides','free sulfur dioxide','total sulfur dioxide', 'density','pH','sulphates','alcohol']
X=pima[feature_cols]
y = pima.quality # Shaun: Good ✅

In [ ]:
X

fixed acidity  volatile acidity  citric acid  ...    pH  sulphates  alcohol
0               7.4             0.700         0.00  ...  3.51       0.56      9.4
1               7.8             0.880         0.00  ...  3.20       0.68      9.8
2               7.8             0.760         0.04  ...  3.26       0.65      9.8
3              11.2             0.280         0.56  ...  3.16       0.58      9.8
4               7.4             0.700         0.00  ...  3.51       0.56      9.4
...             ...               ...          ...  ...   ...        ...      ...
1594            6.2             0.600         0.08  ...  3.45       0.58     10.5
1595            5.9             0.550         0.10  ...  3.52       0.76     11.2
1596            6.3             0.510         0.13  ...  3.42       0.75     11.0
1597            5.9             0.645         0.12  ...  3.57       0.71     10.2
1598            6.0             0.310         0.47  ...  3.39       0.66     11.0

[1599 rows x 11 columns]

In [ ]:
y

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

## Select a training feature


In [ ]:

#for i in range(0,11)
#X1=X.iloc[:,i:i+1]
X1=X.iloc[:,0:1] # Shaun: Good ✅
X2=X.iloc[:,1:2]
X3=X.iloc[:,2:3]
X4=X.iloc[:,3:4]
X5=X.iloc[:,4:5]
X6=X.iloc[:,5:6]
X7=X.iloc[:,6:7]
X8=X.iloc[:,7:8]
X9=X.iloc[:,8:9]
X10=X.iloc[:,9:10]
X11=X.iloc[:,10:11]

## Check shape and data

In [ ]:
X1

fixed acidity
0               7.4
1               7.8
2               7.8
3              11.2
4               7.4
...             ...
1594            6.2
1595            5.9
1596            6.3
1597            5.9
1598            6.0

[1599 rows x 1 columns]

In [ ]:
X1.shape # Shaun: Good ✅ I don't see where you check the shape of y's

(1599, 1)

In [ ]:
X2

volatile acidity
0                0.700
1                0.880
2                0.760
3                0.280
4                0.700
...                ...
1594             0.600
1595             0.550
1596             0.510
1597             0.645
1598             0.310

[1599 rows x 1 columns]

In [ ]:
X2.shape

(1599, 1)

In [ ]:
X3

citric acid
0            0.00
1            0.00
2            0.04
3            0.56
4            0.00
...           ...
1594         0.08
1595         0.10
1596         0.13
1597         0.12
1598         0.47

[1599 rows x 1 columns]

In [ ]:
X3.shape

(1599, 1)

In [ ]:
X4

residual sugar
0                1.9
1                2.6
2                2.3
3                1.9
4                1.9
...              ...
1594             2.0
1595             2.2
1596             2.3
1597             2.0
1598             3.6

[1599 rows x 1 columns]

In [ ]:
X5

chlorides
0         0.076
1         0.098
2         0.092
3         0.075
4         0.076
...         ...
1594      0.090
1595      0.062
1596      0.076
1597      0.075
1598      0.067

[1599 rows x 1 columns]

In [ ]:
X5.shape

(1599, 1)

In [ ]:
X6

free sulfur dioxide
0                    11.0
1                    25.0
2                    15.0
3                    17.0
4                    11.0
...                   ...
1594                 32.0
1595                 39.0
1596                 29.0
1597                 32.0
1598                 18.0

[1599 rows x 1 columns]

In [ ]:
X6.shape

(1599, 1)

In [ ]:
X7

total sulfur dioxide
0                     34.0
1                     67.0
2                     54.0
3                     60.0
4                     34.0
...                    ...
1594                  44.0
1595                  51.0
1596                  40.0
1597                  44.0
1598                  42.0

[1599 rows x 1 columns]

In [ ]:
X7.shape

(1599, 1)

In [ ]:
X8

density
0     0.99780
1     0.99680
2     0.99700
3     0.99800
4     0.99780
...       ...
1594  0.99490
1595  0.99512
1596  0.99574
1597  0.99547
1598  0.99549

[1599 rows x 1 columns]

In [ ]:
X8.shape

(1599, 1)

In [ ]:
X9

pH
0     3.51
1     3.20
2     3.26
3     3.16
4     3.51
...    ...
1594  3.45
1595  3.52
1596  3.42
1597  3.57
1598  3.39

[1599 rows x 1 columns]

In [ ]:
X9.shape

(1599, 1)

In [ ]:
X10

sulphates
0          0.56
1          0.68
2          0.65
3          0.58
4          0.56
...         ...
1594       0.58
1595       0.76
1596       0.75
1597       0.71
1598       0.66

[1599 rows x 1 columns]

In [ ]:
X10.shape

(1599, 1)

In [ ]:
X11

alcohol
0         9.4
1         9.8
2         9.8
3         9.8
4         9.4
...       ...
1594     10.5
1595     11.2
1596     11.0
1597     10.2
1598     11.0

[1599 rows x 1 columns]

In [ ]:
X11.shape

(1599, 1)

In [ ]:
y.shape # Shaun: Good ✅

(1599,)

## Split into training and testing

In [ ]:
X_train1, X_test1, y_train1, y_test1 = train_test_split(X1, y, test_size=0.2, random_state=4271)
X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y, test_size=0.2, random_state=4271)
X_train3, X_test3, y_train3, y_test3 = train_test_split(X3, y, test_size=0.2, random_state=4271)
X_train4, X_test4, y_train4, y_test4 = train_test_split(X4, y, test_size=0.2, random_state=4271)
X_train5, X_test5, y_train5, y_test5 = train_test_split(X5, y, test_size=0.2, random_state=4271)
X_train6, X_test6, y_train6, y_test6 = train_test_split(X6, y, test_size=0.2, random_state=4271)
X_train7, X_test7, y_train7, y_test7 = train_test_split(X7, y, test_size=0.2, random_state=4271)
X_train8, X_test8, y_train8, y_test8 = train_test_split(X8, y, test_size=0.2, random_state=4271)
X_train9, X_test9, y_train9, y_test9= train_test_split(X9, y, test_size=0.2, random_state=4271)
X_train10, X_test10, y_train10, y_test10 = train_test_split(X10, y, test_size=0.2, random_state=4271)
X_train11, X_test11, y_train11, y_test11 = train_test_split(X11, y, test_size=0.2, random_state=4271)

# Shaun: Great ✅✅✅✅✅ I don't know why you did this for ALL features when you were told to just pick one.

## Check the training features shape

In [ ]:
X_train1.shape # Shaun: Good ✅

(1279, 1)

In [ ]:
X_train2.shape

(1279, 1)

In [ ]:
X_train3.shape

(1279, 1)

In [ ]:
X_train4.shape

(1279, 1)

In [ ]:
X_train5.shape

(1279, 1)

In [ ]:
X_train6.shape

(1279, 1)

In [ ]:
X_train7.shape

(1279, 1)

In [ ]:
X_train8.shape

(1279, 1)

In [ ]:
X_train9.shape

(1279, 1)

In [ ]:
X_train10.shape

(1279, 1)

In [ ]:
X_train11.shape

(1279, 1)

## Check the testing features shape

In [ ]:
X_test1.shape # Shaun: Good ✅

(320, 1)

In [ ]:
X_test2.shape

(320, 1)

In [ ]:
X_test3.shape

(320, 1)

In [ ]:
X_test4.shape

(320, 1)

In [ ]:
X_test5.shape

(320, 1)

In [ ]:
X_test6.shape

(320, 1)

In [ ]:
X_test7.shape

(320, 1)

In [ ]:
X_test8.shape

(320, 1)

In [ ]:
X_test9.shape

(320, 1)

In [ ]:
X_test10.shape

(320, 1)

In [ ]:
X_test11.shape

(320, 1)

## Define a linear regression model

In [ ]:
model1 = LinearRegression() # Shaun: Good ✅
model2 = LinearRegression()
model3 = LinearRegression()
model4 = LinearRegression()
model5 = LinearRegression()
model6 = LinearRegression()
model7 = LinearRegression()
model8 = LinearRegression()
model9 = LinearRegression()
model10 = LinearRegression()
model11 = LinearRegression()

## Look at what has just being initialised

In [ ]:
model1

LinearRegression()

In [ ]:
model2

LinearRegression()

In [ ]:
model3

LinearRegression()

In [ ]:
model4

LinearRegression()

In [ ]:
model5

LinearRegression()

In [ ]:
model6

LinearRegression()

In [ ]:
model7

LinearRegression()

In [ ]:
model8

LinearRegression()

In [ ]:
model9

LinearRegression()

In [ ]:
model10

LinearRegression()

In [ ]:
model11

LinearRegression()

## Train the model/fit the model

In [ ]:
model1.fit(X_train1, y_train1) # Shaun: Good ✅
model2.fit(X_train2, y_train2)
model3.fit(X_train3, y_train3)
model4.fit(X_train4, y_train4)
model5.fit(X_train5, y_train5)
model6.fit(X_train6, y_train6)
model7.fit(X_train7, y_train7)
model8.fit(X_train8, y_train8)
model9.fit(X_train9, y_train9)
model10.fit(X_train10, y_train10)
model11.fit(X_train11, y_train11)


LinearRegression()

## Print the coefficient and intercept for each model

In [ ]:
model1.coef_

array([0.06192035])

In [ ]:
model1.intercept_

5.1309364866828435

## The model1

The model, given the above output, is as follows:

y = 5 + 0.06x

y = 5 + 0.06d1, d1 = fixed acidity

y = 5 d0 + 0.06 d1, d0 = 1, d1 = fixed acidity
# Shaun: Very good ✅

In [ ]:
model2.coef_ # Shaun: Good ✅

array([-1.75938164])

In [ ]:
model2.intercept_ # Shaun: Good ✅

6.573351140936321

## The model2

The model2, given the above output, is as follows:

y = 6.6 -1.7x

y = 6.6 -1.7d1, d1 = volatile acidity

y = 6.6 d0 - 1.7d1, d0 = 1, d1 = volatile acidity

In [ ]:
model3.coef_

array([0.94897999])

In [ ]:
model3.intercept_

5.388102043072222

## The model3

The model3, given the above output, is as follows:

y = 5 + 0.9x

y = 5 + 0.9d1, d1 = citric acid

y = 5 d0 + 0.9 d1, d0 = 1, d1 = citric acid

In [ ]:
model4.coef_

array([0.01259182])

In [ ]:
model4.intercept_

5.615971779185108

## The model4

The model, given the above output, is as follows:

y = 6 + 0.01x

y = 6 + 0.01d1, d1 = residual sugar

y = 6 d0 + 0.01 d1, d0 = 1, d1 = residual sugar

In [ ]:
model5.coef_

array([-1.96591427])

In [ ]:
model5.intercept_

5.820323841860506

## The model5

The model5, given the above output, is as follows:

y = 6 -1.97x

y = 6 -1.97d1, d1 = chlorides

y = 6 d0 - 1.97 d1, d0 = 1, d1 = chlorides

In [ ]:
model6.coef_

array([-0.00511285])

In [ ]:
model6.intercept_

5.729240640315449

## The model6

The model6, given the above output, is as follows:

y = 6 -0.005x

y = 6 0.005d1, d1 = free sulfur dioxide

y = 6 d0 -0.005 d1, d0 = 1, d1 = free sulfur dioxide

In [ ]:
model7.coef_

array([-0.00492622])

In [ ]:
model7.intercept_

5.876158962542957

## The model7

The model7, given the above output, is as follows:

y = 6 -0.005x

y = 6 -0.005d1, d1 = total sulfur dioxie

y = 6 d0 -0.005 d1, d0 = 1, d1 = total sulfur dioxie

In [ ]:
model8.coef_

array([-65.07191036])

In [ ]:
model8.intercept_

70.5098749488025

## The model8

The model8, given the above output, is as follows:

y = 70 -65x

y = 70 -65d1, d1 = density

y = 70 d0 -65 d1, d0 = 1, d1 = density

In [ ]:
model9.coef_

array([-0.30502641])

In [ ]:
model9.intercept_

6.657008274874933

## The model9

The model9, given the above output, is as follows:

y = 7 -0.3x

y = 7 -0.3 d1, d1 = pH

y = 7 d0 -0.3 d1, d0 = 1, d1 = pH

In [ ]:
model10.coef_

array([1.24365914])

In [ ]:
model10.intercept_

4.8316260136846285

## The model10

The model10, given the above output, is as follows:

y = 5 + 1.2x

y = 5 + 1.2d1, d1 = sulphates

y = 5 d0 + 1.2 d1, d0 = 1, d1 = sulphates

In [ ]:
model11.coef_

array([0.35421218])

In [ ]:
model11.intercept_

1.9556506496755923

## The model11

The model, given the above output, is as follows:

y = 2 + 0.35x

y = 2 + 0.35d1, d1 = alcohol

y = 2 d0 + 0.35 d1, d0 = 1, d1 = alcohol

## Apply the model to the testing features and get the predictions

In [ ]:
Y1_pred = model1.predict(X_test1) # Shaun: Good ✅
Y2_pred = model2.predict(X_test2)
Y3_pred = model3.predict(X_test3)
Y4_pred = model4.predict(X_test4)
Y5_pred = model5.predict(X_test5)
Y6_pred = model6.predict(X_test6)
Y7_pred = model7.predict(X_test7)
Y8_pred = model8.predict(X_test8)
Y9_pred = model9.predict(X_test9)
Y10_pred = model10.predict(X_test10)
Y11_pred = model11.predict(X_test11)

## Determine the performance of the model on the testing data( for each model) using mean absolute error

In [ ]:
mean_absolute_error(y_test1, Y1_pred) # Shaun: Good ✅

0.6903580824803222

In [ ]:
mean_absolute_error(y_test2, Y2_pred)

0.604497020388465

In [ ]:
mean_absolute_error(y_test3, Y3_pred)

0.6756674057349019

In [ ]:
mean_absolute_error(y_test4, Y4_pred)

0.6903330200977364

In [ ]:
mean_absolute_error(y_test5, Y5_pred)

0.675432919932254

In [ ]:
mean_absolute_error(y_test6, Y6_pred)

0.6891922459980284

In [ ]:
mean_absolute_error(y_test7, Y7_pred)

0.6609489061410093

In [ ]:
mean_absolute_error(y_test8, Y8_pred)

0.6572618966526579

In [ ]:
mean_absolute_error(y_test9, Y9_pred)

0.6907333963567034

In [ ]:
mean_absolute_error(y_test10, Y10_pred)

0.6498959511288388

In [ ]:
mean_absolute_error(y_test11, Y11_pred)

0.5596797413653187

alcohol model has the best mean absolute error

## Part 2

## Multivariant regession

## Create X and y variables (features and target)

In [ ]:
feature_cols =['fixed acidity','volatile acidity','citric acid','residual sugar', 'chlorides','free sulfur dioxide','total sulfur dioxide', 'density','pH','sulphates','alcohol']
X=pima[feature_cols] # Shaun: Good ✅
y = pima.quality # Shaun: Good ✅

In [ ]:
X

fixed acidity  volatile acidity  citric acid  ...    pH  sulphates  alcohol
0               7.4             0.700         0.00  ...  3.51       0.56      9.4
1               7.8             0.880         0.00  ...  3.20       0.68      9.8
2               7.8             0.760         0.04  ...  3.26       0.65      9.8
3              11.2             0.280         0.56  ...  3.16       0.58      9.8
4               7.4             0.700         0.00  ...  3.51       0.56      9.4
...             ...               ...          ...  ...   ...        ...      ...
1594            6.2             0.600         0.08  ...  3.45       0.58     10.5
1595            5.9             0.550         0.10  ...  3.52       0.76     11.2
1596            6.3             0.510         0.13  ...  3.42       0.75     11.0
1597            5.9             0.645         0.12  ...  3.57       0.71     10.2
1598            6.0             0.310         0.47  ...  3.39       0.66     11.0

[1599 rows x 11 columns]

In [ ]:
y

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

## Check the shape!

In [ ]:
X.shape # Shaun: Good ✅

(1599, 11)

In [ ]:
y.shape # Shaun: Good ✅

(1599,)

## Split the data into training and testing sets

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Shaun: Great ✅✅✅✅✅

## Check the training features shape

In [ ]:
X_train.shape # Shaun: Good ✅

(1279, 11)

## Check the testing features shape

In [ ]:
X_test.shape # Shaun: Good ✅ Where are the checking of the shapes of the y's?

(320, 11)

## Define a linear regression model

In [ ]:
model = LinearRegression() # Shaun: Good ✅

## Look at what has just being initialised

In [ ]:
model

LinearRegression()

## Train the model/fit the model

In [ ]:
model.fit(X_train, y_train) # Shaun: Good ✅

LinearRegression()

## Print the coefficients

In [ ]:
model.coef_

array([ 2.30853339e-02, -1.00130443e+00, -1.40821461e-01,  6.56431104e-03,
       -1.80650315e+00,  5.62733439e-03, -3.64444893e-03, -1.03515936e+01,
       -3.93687732e-01,  8.41171623e-01,  2.81889567e-01])

## Print the intercept

In [ ]:
model.intercept_

14.355105195764825

## The model.
y= 14 + 0.023d1 -1d2 -0.14d3 + 0.0066d4 -1.8d5 + 0.0056d6 -0.0036d7 -0.1d8 -0.39d9 + 0.84d10 + 0.28d11

## Apply the model to the testing features and get the predictions

In [ ]:
Y_pred = model.predict(X_test) # Shaun: Good ✅

### Mean absolute error

In [ ]:
mean_absolute_error(y_test, Y_pred) # Shaun: Good ✅

0.5035304415524372

### # Shaun: Where are your conclusions to which model is best and why?